### PR TITLE
weightDescriptors supported and IO/initializer distinction made clear

### PR DESCRIPTION
**Pull request by varunjain99**
Support for weight descriptors added.
Removed bug that relied on the order of inputs to execute the computation.
Distinction between parameters and constants on the XLA side (before initializers were incorrectly made parameters).
Should now be able to run graph with different inputs.

### CODE REVIEW

**Comment by varunjain99** [comment]
@bddppq @yinghai

File: onnx_xla/backend.cc [hunk 1]
```diff
@@ -101,6 +101,28 @@ namespace onnx_xla {
     #undef OPERATION
   }
 
+  std::unique_ptr<Literal> XlaExecutor::descriptorToLiteral(const onnxTensorDescriptor& t) {
+    std::vector<int64> sizes;
```
Comment (Maratyszcza) [inline onnx_xla/backend.cc]: `std::vector<int64> sizes(&t.shape[0], &t.shape[t.dimensions])` will get you a copy of `t.shape` in `std::vector`

File: onnx_xla/backend.cc [hunk 2]
```diff
@@ -101,6 +101,28 @@ namespace onnx_xla {
...
+      auto l = std::unique_ptr<Literal>(new Literal(ShapeUtil::MakeShape(      \
+               NativeToPrimitiveType<type_to>(), sizes)));                     \
+      tensorflow::gtl::MutableArraySlice<type_to> l_data = l->data<type_to>(); \
+      type_from* inputData = (type_from*) t.buffer;                            \
+      for (auto i = 0; i < num_elements; ++i) {                                \
+        l_data[i] = (type_to) inputData[i];                                    \
+      }                                                                        \
+      return l;                                                                \
+
+    SWITCH(onnxifiToOnnx(t.dataType))
```
Comment (Maratyszcza) [inline onnx_xla/backend.cc]: This is very cryptic. Any way to avoid macros?

File: onnx_xla/backend_test.cc [hunk 3]
```diff
@@ -48,11 +48,11 @@ namespace onnx_xla  {
     output.buffer = (onnxPointer) new float[24];
    
      //Execute using XLA backend
-    XlaTransform runner(std::move(relu_graph), "relu");
+    XlaTransform runner(std::move(relu_graph), "relu", 0, NULL);
```
Comment (Maratyszcza) [inline onnx_xla/backend_test.cc]: `NULL` -> `nullptr`

File: onnx_xla/onnx_xla_interface.cc [hunk 4]
```diff
@@ -208,7 +210,8 @@ ONNXIFI_PUBLIC ONNXIFI_CHECK_RESULT onnxStatus ONNXIFI_ABI ONNXIFI_SYMBOL_NAME(
       return ONNXIFI_STATUS_INVALID_SIZE;
     }
 
-    *graph = (onnxGraph) backendcontroller->build(onnxModel, onnxModelSize);
+    *graph = (onnxGraph) backendcontroller->build(onnxModel, onnxModelSize,
```
Comment (Maratyszcza) [inline onnx_xla/onnx_xla_interface.cc]: `*graph` must be set to `NULL`/`nullptr` if the function fails. I suggest to do it in the first line of the function, and then overwrite null value if function succeeds.

File: onnx_xla/backend.cc [hunk 5]
```diff
@@ -187,37 +203,69 @@ namespace onnx_xla {
     value_to_op_[v] = std::move(builder_.GetTupleElement(op, index));
   }
 
-  void XlaTransform::fillIOMetadata()  {
-    executor_->num_inputs_ = (uint32_t) ((int64_t) (ir_->inputs().size()) - (int64_t) (ir_->initializers().size()));
-    executor_->num_outputs_ = (uint32_t) ir_->outputs().size();
+  void XlaTransform::handleInputs()  {
+    ONNX_ASSERT(ir_->initializers().size() == 0 || !weight_descriptors_);
```
Comment (Maratyszcza) [inline onnx_xla/backend.cc]: If this condition fails, ONNXIFI implementation must return an error, not crash the whole process via `abort`

File: onnx_xla/backend.cc [hunk 6]
```diff
@@ -187,37 +203,69 @@ namespace onnx_xla {
...
-      isInitialized[s] = true;
+
+    if (weight_descriptors_)  {
+      executor_->num_inputs_ = (uint32_t) ((int64_t) (ir_->inputs().size()) - (int64_t) (weights_count_));
+      for (auto i = 0; i < weights_count_; ++i)  {
+        std::string name(weight_descriptors_[i].name);
+        isInitialized[name] = true;
+        const onnxTensorDescriptor& t = weight_descriptors_[i];
+        const Value* v = inputNameToValue[name];
+        ONNX_ASSERT(onnxifiToOnnx(t.dataType) == v->elemType());
```
Comment (Maratyszcza) [inline onnx_xla/backend.cc]: Same

File: onnx_xla/backend.cc (hunk 2 above)
Comment (bddppq) [reply onnx_xla/backend.cc]: Looks to me one macro for doing type switch is unavoidable, but we can use template function for `OPERATION` and let the type switch macro do the type dispatch.

```
#define dispatch(data_type, operation, ...) \
  swtich(data_type) {
  case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
    return operation<float, float>(__VA_ARGS__);
  }
  ...
  }

template<typename From, typename To>
std::unique_ptr<Literal> XlaExecutor::tensorToLiteralImpl(const Tensor& t) {
  // access the underlying data ptr via t.<From>data(), this needs to be added to the onnx Tensor class in onnx/common/tensor.h, similar to here https://github.com/pytorch/pytorch/blob/master/caffe2/core/tensor.h#L492-L506
  ...
}

std::unique_ptr<Literal> XlaExecutor::tensorToLiteral(const Tensor& t) {
  return dispatch(t.dataType, tensorToLiteralImpl, t);
}

template<typename From typename To>
std::unique_ptr<Literal> XlaExecutor::descriptorToLiteralImpl(const onnxTensorDescriptor& t) {
  ...
}

std::unique_ptr<Literal> XlaExecutor::descriptorToLiteral(const onnxTensorDescriptor& t) {
  return dispatch(t.dataType, descriptorToLiteralImpl, t);
}

}
```

functionality-wise it's same, just wanna make it more organized.

File: onnx_xla/onnx_xla_interface.cc (hunk 4 above)
Comment (varunjain99) [reply onnx_xla/onnx_xla_interface.cc]: added *graph = NULL to beginning of try block

File: onnx_xla/backend_test.cc (hunk 3 above)
Comment (varunjain99) [reply onnx_xla/backend_test.cc]: changed

File: onnx_xla/backend.cc (hunk 5 above)
Comment (varunjain99) [reply onnx_xla/backend.cc]: support for returning status added

File: onnx_xla/backend.cc (hunk 6 above)
Comment (varunjain99) [reply onnx_xla/backend.cc]: support for returning status added

File: onnx_xla/backend.cc (hunk 2 above)
Comment (varunjain99) [reply onnx_xla/backend.cc]: Will refactor later in separate PR - once functionality added to onnx

**Pull request merged**